### PR TITLE
Mock alu plot

### PR DESCRIPTION
**Pull request by oharboe**
I wanted to understand area and frequency of the different operations as well as the performance when all operations are combined.

image

### CODE REVIEW

File: flow/Makefile
Comment (maliberty) [inline flow/Makefile:1]: Since you mixed the other PR into this one I can't approve it as is.

**Comment by maliberty** [comment]
What does it mean that various operations appear more than once in your plot?

**Comment by oharboe** [comment]
> What does it mean that various operations appear more than once in your plot?

Each plot is an ALU implementation that has that group of operations.

I'm interested in seeing what happens if I have all operations, less multiply, or only the barrel shifter, etc.

File: flow/Makefile
Comment (oharboe) [reply flow/Makefile:1]: No problem. I will update and rebase this on #1377  when it is cooked. #1377 needs to be decided on first.

**Comment by maliberty** [comment]
Its odd the MULT is faster than ADD.

**Comment by oharboe** [comment]
> Its odd the MULT is faster than ADD.

Indeed. Merits further investigation...

**Comment by oharboe** [comment]
@maliberty Looked at a 2 bit ADD only ALU. I don't see anything that's obviously out of place.

```
module MockAlu(
  input        clock,
  input        reset,
  input  [3:0] io_op,
  input  [1:0] io_a,
  input  [1:0] io_b,
  output [1:0] io_out
);
  reg [1:0] a;
  reg [1:0] b;
  wire [2:0] _extendedResult_T = a + b;
  wire [3:0] _extendedResult_T_1 = {{1'd0}, _extendedResult_T};
  wire [2:0] extendedResult = _extendedResult_T_1[2:0];
  reg [1:0] io_out_REG;
  assign io_out = io_out_REG;
  always @(posedge clock) begin
    a <= io_a;
    b <= io_b;
    io_out_REG <= extendedResult[1:0];
  end
endmodule
```

1_synth.v:

```
module MockAlu(clock, reset, io_op, io_a, io_b, io_out);
  wire _00_;
  wire _01_;
  wire _02_;
  wire _03_;
  wire _04_;
  wire _05_;
  wire _06_;
  wire _07_;
  wire _08_;
  wire _09_;
  wire \_extendedResult_T[0] ;
  wire \_extendedResult_T[1] ;
  wire \a[0] ;
  wire \b[0] ;
  input clock;
  input [1:0] io_a;
  input [1:0] io_b;
  input [3:0] io_op;
  output [1:0] io_out;
  input reset;
  XOR2x1_ASAP7_75t_R _10_ (
    .A(_04_),
    .B(_06_),
    .Y(_09_)
  );
  XNOR2x1_ASAP7_75t_R _11_ (
    .A(_02_),
    .B(_09_),
    .Y(\_extendedResult_T[1] )
  );
  INVx1_ASAP7_75t_R _12_ (
    .A(clock),
    .Y(_08_)
  );
  INVx1_ASAP7_75t_R _13_ (
    .A(_00_),
    .Y(io_out[1])
  );
  INVx1_ASAP7_75t_R _14_ (
    .A(_01_),
    .Y(io_out[0])
  );
  INVx1_ASAP7_75t_R _15_ (
    .A(_03_),
    .Y(\b[0] )
  );
  INVx1_ASAP7_75t_R _16_ (
    .A(_05_),
    .Y(\a[0] )
  );
  INVx1_ASAP7_75t_R _17_ (
    .A(_07_),
    .Y(\_extendedResult_T[0] )
  );
  HAxp5_ASAP7_75t_R _18_ (
    .A(\a[0] ),
    .B(\b[0] ),
    .CON(_06_),
    .SN(_07_)
  );
  DFFLQNx1_ASAP7_75t_R _19_ (
    .CLK(_08_),
    .D(io_a[0]),
    .QN(_05_)
  );
  DFFLQNx1_ASAP7_75t_R _20_ (
    .CLK(_08_),
    .D(io_a[1]),
    .QN(_04_)
  );
  DFFLQNx1_ASAP7_75t_R _21_ (
    .CLK(_08_),
    .D(io_b[0]),
    .QN(_03_)
  );
  DFFLQNx1_ASAP7_75t_R _22_ (
    .CLK(_08_),
    .D(io_b[1]),
    .QN(_02_)
  );
  DFFLQNx1_ASAP7_75t_R _23_ (
    .CLK(_08_),
    .D(\_extendedResult_T[0] ),
    .QN(_01_)
  );
  DFFLQNx1_ASAP7_75t_R _24_ (
    .CLK(_08_),
    .D(\_extendedResult_T[1] ),
    .QN(_00_)
  );
endmodule
```

```
>>> report_clock_min_period
clock period_min = 121.80 fmax = 8209.85
>>> report_checks -path_delay max
Startpoint: _23_ (falling edge-triggered flip-flop clocked by clock')
Endpoint: io_out[0] (output port clocked by clock)
Path Group: clock
Path Type: max

  Delay    Time   Description
---------------------------------------------------------
   0.00    0.00   clock clock' (fall edge)
  38.00   38.00   clock network delay (propagated)
   0.00   38.00 v _23_/CLK (DFFLQNx3_ASAP7_75t_R)
  45.31   83.31 v _23_/QN (DFFLQNx3_ASAP7_75t_R)
   7.11   90.42 ^ _14_/Y (INVx1_ASAP7_75t_R)
  11.71  102.13 ^ output5/Y (BUFx2_ASAP7_75t_R)
   0.01  102.14 ^ io_out[0] (out)
         102.14   data arrival time

 750.00  750.00   clock clock (rise edge)
   0.00  750.00   clock network delay (propagated)
   0.00  750.00   clock reconvergence pessimism
-150.00  600.00   output external delay
         600.00   data required time
---------------------------------------------------------
         600.00   data required time
        -102.14   data arrival time
---------------------------------------------------------
         497.86   slack (MET)
```

**Comment by oharboe** [comment]
I've adjusted the MockAlu.scala so that there's no mux(it was setting the result 0 when io.op is not ADD) when there's only an ADD, then I get:

```
$ echo report_clock_min_period | make open_final
[deleted]openroad> report_clock_min_period
clock period_min = 374.36 fmax = 2671.23
openroad> 
$ echo report_clock_min_period | make open_cts
[deleted]
clock period_min = 403.61 fmax = 2477.61
openroad> 
```

image

**Comment by oharboe** [comment]
@maliberty Ready to merge for my part.

Note! I did a force push to remove dependency on the upcoming feature of #1385.

#1385 is only needed to generate the plots, so this PR can go in before or after that feature.

**Comment by oharboe** [comment]
Plotting 8, 16, 32 and 64 bit ADD:
image

**Comment by oharboe** [comment]
Various multiplication algorithms and multiply and accumulate.

interestingly the MAC takes more area, but is otherwise the same speed. The spread in area is quite small... also speed, except for the ripple algorithm. Interesting...

image

**Review by maliberty — APPROVED** [review]

**Pull request merged**